File: app/src/main/java/wxm/KeepAccount/Base/db/IncomeNoteItem.java
package wxm.KeepAccount.Base.db;

import android.os.Parcel;
import android.os.Parcelable;

import com.j256.ormlite.field.DataType;
import com.j256.ormlite.field.DatabaseField;
import com.j256.ormlite.table.DatabaseTable;

import java.math.BigDecimal;
import java.sql.Timestamp;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import wxm.KeepAccount.Base.data.AppGobalDef;


/**
 * 支出记录数据
 * Created by 123 on 2016/5/3.
 */
@DatabaseTable(tableName = "tbIncomeNote")
public class IncomeNoteItem implements Parcelable {
    public final static String FIELD_TS         = "ts";
    public final static String FIELD_USR        = "usr_id";

    @DatabaseField(generatedId = true, columnName = "_id", dataType = DataType.INTEGER)
    private int _id;

    @DatabaseField(columnName = "usr_id", foreign = true, foreignColumnName = UsrItem.FIELD_ID,
            canBeNull = false)
    private UsrItem usr;

    @DatabaseField(columnName = "budget_id", foreign = true, foreignColumnName = BudgetItem.FIELD_ID,
            canBeNull = true)
    private BudgetItem budget;

    @DatabaseField(columnName = "info", canBeNull = false, dataType = DataType.STRING)
    private String info;

    @DatabaseField(columnName = "note", dataType = DataType.STRING)
    private String note;

    @DatabaseField(columnName = "val", dataType = DataType.BIG_DECIMAL)
    private BigDecimal val;

    @DatabaseField(columnName = "ts", dataType = DataType.TIME_STAMP)
    private Timestamp ts;

    public int getId() {
        return _id;
    }

    public void setId(int _id) {
        this._id = _id;
    }

    public String getInfo() {
        return info;
    }

    public void setInfo(String info) {
        this.info = info;
    }

    public String getNote() {
        return note;
    }

    public void setNote(String note) {
        this.note = note;
    }

    public BigDecimal getVal() {
        return val;
    }

    public void setVal(BigDecimal val) {
        this.val = val;
    }

    public Timestamp getTs() {
        return ts;
    }

    public void setTs(Timestamp tsval) {
        this.ts = tsval;
    }

    public UsrItem getUsr() {
        return usr;
    }

    public void setUsr(UsrItem usr) {
        this.usr = usr;
    }


    public BudgetItem getBudget() {
        return budget;
    }

    public void setBudget(BudgetItem budget) {
        this.budget = budget;
    }


    public IncomeNoteItem()
    {
        setTs(new Timestamp(0));
        setVal(BigDecimal.ZERO);
        setInfo("");
        setNote("");
    }

    @Override
    public String toString()
    {
        return String.format(Locale.CHINA
                ,"info : %s, val : %f, timestamp : %s\nnote : %s"
                ,getInfo() ,getVal() ,getTs().toString() ,getNote());
    }

    public int describeContents() {
        return 0;
    }

    public void writeToParcel(Parcel out, int flags) {
        out.writeInt(getId());

        int usrid = AppGobalDef.INVALID_ID;
        UsrItem ui = getUsr();
        if(null != ui)
            usrid = ui.getId();
        out.writeInt(usrid);

        int budgetid = AppGobalDef.INVALID_ID;
        BudgetItem bi = getBudget();
        if(null != bi)
            budgetid = bi.get_id();
        out.writeInt(budgetid);

        out.writeString(getInfo());
        out.writeString(getNote());
        out.writeString(getVal().toString());
        out.writeString(getTs().toString());
    }

    public static final Parcelable.Creator<IncomeNoteItem> CREATOR
            = new Parcelable.Creator<IncomeNoteItem>() {
        public IncomeNoteItem createFromParcel(Parcel in) {
            return new IncomeNoteItem(in);
        }

        public IncomeNoteItem[] newArray(int size) {
            return new IncomeNoteItem[size];
        }
    };

    private IncomeNoteItem(Parcel in)   {
        setId(in.readInt());

        setUsr(new UsrItem());
        getUsr().setId(in.readInt());

        setBudget(new BudgetItem());
        getBudget().set_id(in.readInt());

        setInfo(in.readString());
        setNote(in.readString());
        setVal(new BigDecimal(in.readString()));

        try {
            setTs(new Timestamp(0));

            SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US);
            Date date = format.parse(in.readString());
            getTs().setTime(date.getTime());
        }
        catch (ParseException ex)
        {
            setTs(new Timestamp(0));
        }
    }
}

File: app/src/main/java/wxm/KeepAccount/item/IncomeNoteItem.kt
package wxm.KeepAccount.item

import com.j256.ormlite.field.DataType
import com.j256.ormlite.field.DatabaseField
import com.j256.ormlite.table.DatabaseTable
import wxm.KeepAccount.define.GlobalDef
import wxm.KeepAccount.define.IPublicClone
import wxm.KeepAccount.improve.toMoneyStr
import wxm.androidutil.db.IDBRow

import java.math.BigDecimal
import java.sql.Timestamp
import java.util.*


/**
 * income record
 * Created by WangXM on 2016/5/3.
 */
@DatabaseTable(tableName = "tbIncomeNote")
class IncomeNoteItem(override var tag: Any? = null) :
        INote, IImage, IDBRow<Int>, Cloneable, IPublicClone {
    @DatabaseField(generatedId = true, columnName = INote.FILELD_ID, dataType = DataType.INTEGER)
    override var id: Int = GlobalDef.INVALID_ID

    @DatabaseField(columnName = INote.FIELD_USR, foreign = true, foreignColumnName = UsrItem.FIELD_ID, canBeNull = false)
    override var usr: UsrItem? = null

    @DatabaseField(columnName = INote.FIELD_INFO, canBeNull = false, dataType = DataType.STRING)
    override var info: String = ""

    @DatabaseField(columnName = INote.FIELD_NOTE, dataType = DataType.STRING)
    override var note: String? = null

    @DatabaseField(columnName = INote.FIELD_AMOUNT, dataType = DataType.BIG_DECIMAL)
    override var amount: BigDecimal = BigDecimal.ZERO

    @DatabaseField(columnName = INote.FIELD_TS, dataType = DataType.TIME_STAMP)
    override var ts: Timestamp = Timestamp(0)

    /// IImage START
    override val holderId
        get() = id

    override val holderType
        get() = noteType()

    override var images: LinkedList<NoteImageItem> = LinkedList()
        set(value) {
            field.clear()
            field.addAll(value)
        }
    /// IImage END

    override fun noteType(): String = GlobalDef.STR_RECORD_INCOME

    override fun toString(): String {
        return String.format(Locale.CHINA,
                "info : %s, amount : %f, timestamp : %s\nnote : %s",
                info, amount, ts.toString(), note)
    }

    override fun getID(): Int? {
        return id
    }

    override fun setID(integer: Int) {
        id = integer
    }

    override fun clone(): Any {
        val obj = IncomeNoteItem(tag)

        obj.id = this.id
        this.usr?.let {
            obj.usr = it.publicClone() as UsrItem
        }

        obj.amount = this.amount
        obj.info = this.info
        obj.note = this.note
        obj.ts = this.ts

        obj.images = this.images
        return obj
    }

    override fun publicClone(): Any {
        return clone()
    }
}

